File: Sources/PetsGraphCore/PlaybackTimeline.swift
import Foundation

public struct TimelineSample: Equatable, Sendable {
  public let segmentIndex: Int
  public let clipID: String
  public let sourceFrameIndex: Int
  public let elapsedSeconds: TimeInterval
  public let rootMotionXPt: Double

  public init(
    segmentIndex: Int,
    clipID: String,
    sourceFrameIndex: Int,
    elapsedSeconds: TimeInterval,
    rootMotionXPt: Double
  ) {
    self.segmentIndex = segmentIndex
    self.clipID = clipID
    self.sourceFrameIndex = sourceFrameIndex
    self.elapsedSeconds = elapsedSeconds
    self.rootMotionXPt = rootMotionXPt
  }
}

public struct PlaybackTimeline: Sendable {
  private struct ResolvedFrame: Sendable {
    let sourceFrameIndex: Int
    let startSeconds: TimeInterval
    let durationSeconds: TimeInterval
    let startMotionX: Double
    let endMotionX: Double
  }

  private struct ResolvedSegment: Sendable {
    let clipID: String
    let startSeconds: TimeInterval
    let startMotionX: Double
    let durationSeconds: TimeInterval
    let terminalMotionX: Double
    let repeatForever: Bool
    let frames: [ResolvedFrame]
  }

  private let segments: [ResolvedSegment]

  public let finiteDurationSeconds: TimeInterval
  public let finiteRootMotionXPt: Double

  public init(
    clips: [String: ClipDefinition],
    sequence: DemoSequence
  ) throws {
    var resolved: [ResolvedSegment] = []
    var sequenceTime: TimeInterval = 0
    var sequenceMotion = 0.0

    for segment in sequence.segments {
      guard let clip = clips[segment.clip] else {
        throw PackageValidationError.invalid("timeline references unknown clip \(segment.clip)")
      }
      let built = try Self.buildSegment(
        clip: clip,
        segment: segment,
        sequenceStart: sequenceTime,
        motionStart: sequenceMotion
      )
      resolved.append(built)
      if !segment.repeatForever {
        sequenceTime += built.durationSeconds
        sequenceMotion += built.terminalMotionX
      }
    }

    segments = resolved
    finiteDurationSeconds = sequenceTime
    finiteRootMotionXPt = sequenceMotion
  }

  public func sample(at elapsedSeconds: TimeInterval) -> TimelineSample {
    let elapsed = max(0, elapsedSeconds)
    guard let index = segmentIndex(at: elapsed) else {
      let finalIndex = segments.count - 1
      let final = segments[finalIndex]
      let frame = final.frames.last!
      return TimelineSample(
        segmentIndex: finalIndex,
        clipID: final.clipID,
        sourceFrameIndex: frame.sourceFrameIndex,
        elapsedSeconds: elapsed,
        rootMotionXPt: final.startMotionX + final.terminalMotionX
      )
    }

    let segment = segments[index]
    var local = max(0, elapsed - segment.startSeconds)
    var repeatedMotion = 0.0
    if segment.repeatForever, segment.durationSeconds > 0 {
      let completedCycles = floor(local / segment.durationSeconds)
      repeatedMotion = completedCycles * segment.terminalMotionX
      local.formTruncatingRemainder(dividingBy: segment.durationSeconds)
    } else {
      local = min(local, segment.durationSeconds)
    }

    let frame = frame(in: segment, at: local)
    let frameElapsed = max(0, local - frame.startSeconds)
    let progress = min(1, frameElapsed / frame.durationSeconds)
    let localMotion = frame.startMotionX
      + (frame.endMotionX - frame.startMotionX) * progress
    return TimelineSample(
      segmentIndex: index,
      clipID: segment.clipID,
      sourceFrameIndex: frame.sourceFrameIndex,
      elapsedSeconds: elapsed,
      rootMotionXPt: segment.startMotionX + repeatedMotion + localMotion
    )
  }

  public func clipIDsNear(segmentIndex: Int, lookahead: Int = 2) -> [String] {
    guard segments.indices.contains(segmentIndex) else {
      return []
    }
    let end = min(segments.count - 1, segmentIndex + max(0, lookahead))
    var seen = Set<String>()
    return segments[segmentIndex...end].compactMap { segment in
      seen.insert(segment.clipID).inserted ? segment.clipID : nil
    }
  }

  private func segmentIndex(at elapsed: TimeInterval) -> Int? {
    for index in segments.indices {
      let segment = segments[index]
      if segment.repeatForever || elapsed < segment.startSeconds + segment.durationSeconds {
        return index
      }
    }
    return nil
  }

  private func frame(
    in segment: ResolvedSegment,
    at localTime: TimeInterval
  ) -> ResolvedFrame {
    if localTime >= segment.durationSeconds {
      return segment.frames.last!
    }
    var lower = 0
    var upper = segment.frames.count
    while lower < upper {
      let middle = (lower + upper) / 2
      let candidate = segment.frames[middle]
      if candidate.startSeconds + candidate.durationSeconds <= localTime {
        lower = middle + 1
      } else {
        upper = middle
      }
    }
    return segment.frames[min(lower, segment.frames.count - 1)]
  }

  private static func buildSegment(
    clip: ClipDefinition,
    segment: DemoSegment,
    sequenceStart: TimeInterval,
    motionStart: Double
  ) throws -> ResolvedSegment {
    let frameCount = clip.frames.count
    let oneCycleOrder = Array(segment.startFrame..<frameCount)
      + Array(0..<segment.startFrame)
    let order: [Int]
    if let requestedFrameCount = segment.frameCount {
      order = Array(oneCycleOrder.prefix(requestedFrameCount))
    } else {
      order = (0..<segment.cycles).flatMap { _ in oneCycleOrder }
    }

    var frames: [ResolvedFrame] = []
    var localTime: TimeInterval = 0
    var localMotion = 0.0
    for sourceIndex in order {
      let definition = clip.frames[sourceIndex]
      let duration = definition.durationMs / 1_000
      let startX = definition.rootMotionPt[0]
      let endX = sourceIndex + 1 < frameCount
        ? clip.frames[sourceIndex + 1].rootMotionPt[0]
        : clip.rootMotionEndPt[0]
      let delta = endX - startX
      guard delta >= -0.000_001 else {
        throw PackageValidationError.invalid("clip \(clip.id) has negative frame root motion")
      }
      frames.append(
        ResolvedFrame(
          sourceFrameIndex: sourceIndex,
          startSeconds: localTime,
          durationSeconds: duration,
          startMotionX: localMotion,
          endMotionX: localMotion + max(0, delta)
        )
      )
      localTime += duration
      localMotion += max(0, delta)
    }
    guard !frames.isEmpty else {
      throw PackageValidationError.invalid("clip \(clip.id) resolved to no frames")
    }

    return ResolvedSegment(
      clipID: clip.id,
      startSeconds: sequenceStart,
      startMotionX: motionStart,
      durationSeconds: localTime,
      terminalMotionX: localMotion,
      repeatForever: segment.repeatForever,
      frames: frames
    )
  }
}

File: player/windows/src/PetsGraph.Core/PlaybackTimeline.cs
namespace PetsGraph.Core;

public readonly record struct TimelineSample(
    int SegmentIndex,
    string ClipId,
    int SourceFrameIndex,
    double ElapsedSeconds,
    double RootMotionXPt);

public sealed class PlaybackTimeline
{
    private sealed record ResolvedFrame(
        int SourceFrameIndex,
        double StartSeconds,
        double DurationSeconds,
        double StartMotionX,
        double EndMotionX);

    private sealed record ResolvedSegment(
        string ClipId,
        double StartSeconds,
        double StartMotionX,
        double DurationSeconds,
        double TerminalMotionX,
        bool RepeatForever,
        ResolvedFrame[] Frames);

    private readonly ResolvedSegment[] segments;

    public double FiniteDurationSeconds { get; }
    public double FiniteRootMotionXPt { get; }

    public PlaybackTimeline(IReadOnlyDictionary<string, ClipDefinition> clips, DemoSequence sequence)
    {
        var resolved = new List<ResolvedSegment>();
        var sequenceTime = 0.0;
        var sequenceMotion = 0.0;
        foreach (var segment in sequence.Segments)
        {
            if (!clips.TryGetValue(segment.Clip, out var clip))
            {
                throw Invalid($"timeline references unknown clip {segment.Clip}");
            }
            var built = BuildSegment(clip, segment, sequenceTime, sequenceMotion);
            resolved.Add(built);
            if (!segment.RepeatForever)
            {
                sequenceTime += built.DurationSeconds;
                sequenceMotion += built.TerminalMotionX;
            }
        }
        if (resolved.Count == 0)
        {
            throw Invalid("timeline contains no segments");
        }
        segments = [.. resolved];
        FiniteDurationSeconds = sequenceTime;
        FiniteRootMotionXPt = sequenceMotion;
    }

    public TimelineSample Sample(double elapsedSeconds)
    {
        var elapsed = Math.Max(0, elapsedSeconds);
        var index = SegmentIndex(elapsed);
        if (index < 0)
        {
            index = segments.Length - 1;
            var final = segments[index];
            var finalFrame = final.Frames[^1];
            return new(index, final.ClipId, finalFrame.SourceFrameIndex, elapsed,
                final.StartMotionX + final.TerminalMotionX);
        }

        var segment = segments[index];
        var local = Math.Max(0, elapsed - segment.StartSeconds);
        var repeatedMotion = 0.0;
        if (segment.RepeatForever && segment.DurationSeconds > 0)
        {
            var completedCycles = Math.Floor(local / segment.DurationSeconds);
            repeatedMotion = completedCycles * segment.TerminalMotionX;
            local %= segment.DurationSeconds;
        }
        else
        {
            local = Math.Min(local, segment.DurationSeconds);
        }

        var frame = FindFrame(segment, local);
        var progress = Math.Min(1, Math.Max(0, local - frame.StartSeconds) / frame.DurationSeconds);
        var localMotion = frame.StartMotionX + (frame.EndMotionX - frame.StartMotionX) * progress;
        return new(index, segment.ClipId, frame.SourceFrameIndex, elapsed,
            segment.StartMotionX + repeatedMotion + localMotion);
    }

    public IReadOnlyList<string> ClipIdsNear(int segmentIndex, int lookahead = 2)
    {
        if (segmentIndex < 0 || segmentIndex >= segments.Length)
        {
            return [];
        }
        var result = new List<string>();
        var seen = new HashSet<string>(StringComparer.Ordinal);
        var end = Math.Min(segments.Length - 1, segmentIndex + Math.Max(0, lookahead));
        for (var index = segmentIndex; index <= end; index++)
        {
            if (seen.Add(segments[index].ClipId))
            {
                result.Add(segments[index].ClipId);
            }
        }
        return result;
    }

    private int SegmentIndex(double elapsed)
    {
        for (var index = 0; index < segments.Length; index++)
        {
            var segment = segments[index];
            if (segment.RepeatForever || elapsed < segment.StartSeconds + segment.DurationSeconds)
            {
                return index;
            }
        }
        return -1;
    }

    private static ResolvedFrame FindFrame(ResolvedSegment segment, double localTime)
    {
        if (localTime >= segment.DurationSeconds)
        {
            return segment.Frames[^1];
        }
        var lower = 0;
        var upper = segment.Frames.Length;
        while (lower < upper)
        {
            var middle = (lower + upper) / 2;
            var candidate = segment.Frames[middle];
            if (candidate.StartSeconds + candidate.DurationSeconds <= localTime)
            {
                lower = middle + 1;
            }
            else
            {
                upper = middle;
            }
        }
        return segment.Frames[Math.Min(lower, segment.Frames.Length - 1)];
    }

    private static ResolvedSegment BuildSegment(
        ClipDefinition clip,
        DemoSegment segment,
        double sequenceStart,
        double motionStart)
    {
        if (segment.StartFrame < 0 || segment.StartFrame >= clip.Frames.Length || segment.Cycles <= 0)
        {
            throw Invalid($"clip {clip.Id} has an invalid segment range");
        }
        var oneCycle = Enumerable.Range(segment.StartFrame, clip.Frames.Length - segment.StartFrame)
            .Concat(Enumerable.Range(0, segment.StartFrame)).ToArray();
        var order = segment.FrameCount is { } count
            ? oneCycle.Take(count).ToArray()
            : Enumerable.Range(0, segment.Cycles).SelectMany(_ => oneCycle).ToArray();

        var frames = new List<ResolvedFrame>();
        var localTime = 0.0;
        var localMotion = 0.0;
        foreach (var sourceIndex in order)
        {
            var definition = clip.Frames[sourceIndex];
            var duration = definition.DurationMs / 1000.0;
            var startX = definition.RootMotionPt[0];
            var endX = sourceIndex + 1 < clip.Frames.Length
                ? clip.Frames[sourceIndex + 1].RootMotionPt[0]
                : clip.RootMotionEndPt[0];
            var delta = endX - startX;
            if (delta < -0.000001)
            {
                throw Invalid($"clip {clip.Id} has negative frame root motion");
            }
            frames.Add(new(sourceIndex, localTime, duration, localMotion, localMotion + Math.Max(0, delta)));
            localTime += duration;
            localMotion += Math.Max(0, delta);
        }
        if (frames.Count == 0)
        {
            throw Invalid($"clip {clip.Id} resolved to no frames");
        }
        return new(clip.Id, sequenceStart, motionStart, localTime, localMotion, segment.RepeatForever, [.. frames]);
    }

    private static PetPackageValidationException Invalid(string detail) =>
        new($"Invalid pet package: {detail}");
}
